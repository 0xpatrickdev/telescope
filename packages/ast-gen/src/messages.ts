import * as t from '@babel/types';
import { Mutation } from './types';

const tsIdentifier = (name: string, typeAnnotation: t.TSTypeAnnotation) => {
  const el = t.identifier(name);
  el.typeAnnotation = typeAnnotation;
  return el;
}

export const addMsgMethod = ({ methodName, typeUrl, TypeName, methodToCall }) => {
  return t.objectMethod('method', t.identifier(methodName), [
    tsIdentifier('value', t.tsTypeAnnotation(
      t.tsTypeReference(
        t.identifier(TypeName)
      )
    ))
  ], t.blockStatement([
    t.returnStatement(
      t.objectExpression([
        t.objectProperty(
          t.identifier('typeUrl'),
          t.stringLiteral(typeUrl)
        ),
        t.objectProperty(
          t.identifier('value'),
          t.callExpression(
            t.memberExpression(
              t.identifier(TypeName),
              t.identifier(methodToCall)
            ),
            [t.identifier('value')]
          )
        )
      ])
    )]));
};

export const addFromJSONMethod = ({ methodName, typeUrl, TypeName }) => {
  return t.objectMethod('method', t.identifier(methodName), [
    tsIdentifier('value', t.tsTypeAnnotation(
      t.tsTypeReference(
        t.identifier('any')
      )
    ))
  ], t.blockStatement([
    t.returnStatement(
      t.objectExpression([
        t.objectProperty(
          t.identifier('typeUrl'),
          t.stringLiteral(typeUrl)
        ),
        t.objectProperty(
          t.identifier('value'),
          t.callExpression(
            t.memberExpression(
              t.identifier(TypeName),
              t.identifier('fromJSON')
            ),
            [t.identifier('value')]
          )
        )
      ])
    )]));
};

export const addFromPartialMethod = ({ methodName, typeUrl, TypeName }) => {
  return addMsgMethod({ methodName, typeUrl, TypeName, methodToCall: 'fromPartial' });
};

export const addToJSONMethod = ({ methodName, typeUrl, TypeName }) => {
  return addMsgMethod({ methodName, typeUrl, TypeName, methodToCall: 'toJSON' });
};


export const addJsonMethod = ({ methodName, typeUrl, TypeName }) => {
  return t.objectMethod('method', t.identifier(methodName), [
    tsIdentifier('value', t.tsTypeAnnotation(
      t.tsTypeReference(
        t.identifier(TypeName)
      )
    ))
  ], t.blockStatement([
    t.returnStatement(
      t.objectExpression([
        t.objectProperty(
          t.identifier('typeUrl'),
          t.stringLiteral(typeUrl)
        ),
        t.objectProperty(
          t.identifier('value'),
          t.identifier('value'),
          false,
          true
        )
      ])
    )]));
};

export const addEncodedMethod = ({ methodName, typeUrl, TypeName }) => {
  return t.objectMethod('method', t.identifier(methodName), [
    tsIdentifier('value', t.tsTypeAnnotation(
      t.tsTypeReference(
        t.identifier(TypeName)
      )
    ))
  ], t.blockStatement([
    t.returnStatement(
      t.objectExpression([

        t.objectProperty(
          t.identifier('type_url'),
          t.stringLiteral(typeUrl)
        ),

        t.objectProperty(
          t.identifier('value'),
          t.callExpression(
            t.memberExpression(
              t.callExpression(
                t.memberExpression(
                  t.identifier(TypeName),
                  t.identifier('encode')
                ),
                [
                  t.identifier('value')
                ]
              ),
              t.identifier('finish')
            ),
            []
          )
        )
      ])
    )]));
};


export const toObjectWithPartialMethods = (mutations: Mutation[]) => t.exportNamedDeclaration(t.variableDeclaration('const', [
  t.variableDeclarator(t.identifier('messages'), t.objectExpression(
    mutations.map(mutation => addFromPartialMethod(mutation))
  ))]));

export const toObjectWithEncodedMethods = (mutations: Mutation[]) => t.exportNamedDeclaration(t.variableDeclaration('const', [
  t.variableDeclarator(t.identifier('encoded'), t.objectExpression(
    mutations.map(mutation => addEncodedMethod(mutation))
  ))]));

export const toObjectWithJsonMethods = (mutations: Mutation[]) => t.exportNamedDeclaration(t.variableDeclaration('const', [
  t.variableDeclarator(t.identifier('json'), t.objectExpression(
    mutations.map(mutation => addJsonMethod(mutation))
  ))]));

export const toObjectWithToJSONMethods = (mutations: Mutation[]) => t.exportNamedDeclaration(t.variableDeclaration('const', [
  t.variableDeclarator(t.identifier('toJSON'), t.objectExpression(
    mutations.map(mutation => addToJSONMethod(mutation))
  ))]));

export const toObjectWithFromJSONMethods = (mutations: Mutation[]) => t.exportNamedDeclaration(t.variableDeclaration('const', [
  t.variableDeclarator(t.identifier('fromJSON'), t.objectExpression(
    mutations.map(mutation => addFromJSONMethod(mutation))
  ))]));

export const createTypeRegistryObject = (mutation: Mutation) => {
  return t.objectProperty(
    t.stringLiteral(mutation.typeUrl),
    t.identifier(mutation.TypeName)
  );
};

export const createTypeRegistry = (mutations: Mutation[]) => t.exportNamedDeclaration(
  t.variableDeclaration('const', [
    t.variableDeclarator(tsIdentifier(
      'registry',
      t.tsTypeAnnotation(
        t.tsTypeReference(t.identifier('ReadonlyArray'), t.tsTypeParameterInstantiation(
          [
            t.tsTupleType([
              t.tsStringKeyword(),
              t.tsTypeReference(
                t.identifier('GeneratedType')
              )]
            )
          ]
        ))
      )
    ), t.arrayExpression(
      [
        ...mutations.map(mutation => t.arrayExpression(
          [
            t.stringLiteral(mutation.typeUrl),
            t.identifier(mutation.TypeName)
          ]
        ))
      ]
    ))
  ]));

export const createRegistryLoader = () => {
  return t.exportNamedDeclaration(t.variableDeclaration(
    'const',
    [
      t.variableDeclarator(
        t.identifier('load'),
        t.arrowFunctionExpression(
          [
            tsIdentifier('protoRegistry', t.tsTypeAnnotation(
              t.tsTypeReference(
                t.identifier('Registry')
              )
            ))
          ],
          t.blockStatement(
            [
              t.expressionStatement(
                t.callExpression(
                  t.memberExpression(
                    t.callExpression(
                      t.memberExpression(
                        t.identifier(
                          'Object'
                        ),
                        t.identifier('keys')
                      ),
                      [
                        t.identifier('registry')
                      ]
                    ),
                    t.identifier('forEach')
                  ),
                  [
                    t.arrowFunctionExpression(
                      [
                        t.identifier('typeUrl')
                      ],
                      t.blockStatement(
                        [
                          t.expressionStatement(
                            t.callExpression(
                              t.memberExpression(
                                t.identifier(
                                  'protoRegistry'
                                ),
                                t.identifier('register')
                              ),
                              [
                                t.identifier('typeUrl'),
                                t.memberExpression(
                                  t.identifier('registry'),
                                  t.identifier('typeUrl'),
                                  true
                                )
                              ]
                            )
                          )
                        ]
                      )
                    )
                  ]
                )
              )
            ]
          )
        )
      )
    ]
  ))
};