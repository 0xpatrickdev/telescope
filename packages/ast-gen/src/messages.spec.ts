import { addEncodedMethod, addFromPartialMethod, addJsonMethod, addFromJSONMethod, addToJSONMethod, encoded, messages } from './messages';
import generate from '@babel/generator';
import { mutations } from './__fixtures__';
const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}

it('addEncodedMethod', async () => {
    expectCode(addEncodedMethod(mutations[0]));
});

it('addFromPartialMethod', async () => {
    expectCode(addFromPartialMethod(mutations[0]));
});

it('addToJSONMethod', async () => {
    expectCode(addToJSONMethod(mutations[0]));
});

it('addFromJSONMethod', async () => {
    expectCode(addFromJSONMethod(mutations[0]));
});

it('addJsonMethod', async () => {
    expectCode(addJsonMethod(mutations[0]));
});

it('encoded', async () => {
    expectCode(encoded(mutations));
});

it('messages', async () => {
    expectCode(messages(mutations));
});