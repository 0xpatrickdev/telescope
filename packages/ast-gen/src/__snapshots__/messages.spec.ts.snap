// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addEncodedMethod 1`] = `
"joinPool(value: MsgJoinPool) {
  return {
    type_url: \\"/cosmos.pools.transfer.v1.MsgJoinPool\\",
    value: MsgJoinPool.encode(value).finish()
  };
}"
`;

exports[`addFromJSONMethod 1`] = `
"joinPool(value: any) {
  return {
    typeUrl: \\"/cosmos.pools.transfer.v1.MsgJoinPool\\",
    value: MsgJoinPool.fromJSON(value)
  };
}"
`;

exports[`addFromPartialMethod 1`] = `
"joinPool(value: MsgJoinPool) {
  return {
    typeUrl: \\"/cosmos.pools.transfer.v1.MsgJoinPool\\",
    value: MsgJoinPool.fromPartial(value)
  };
}"
`;

exports[`addJsonMethod 1`] = `
"joinPool(value: MsgJoinPool) {
  return {
    typeUrl: \\"/cosmos.pools.transfer.v1.MsgJoinPool\\",
    value
  };
}"
`;

exports[`addToJSONMethod 1`] = `
"joinPool(value: MsgJoinPool) {
  return {
    typeUrl: \\"/cosmos.pools.transfer.v1.MsgJoinPool\\",
    value: MsgJoinPool.toJSON(value)
  };
}"
`;

exports[`createTypeRegistry 1`] = `
"export const registry = {
  \\"/cosmos.pools.transfer.v1.MsgJoinPool\\": MsgJoinPool,
  \\"/cosmos.pools.transfer.v1.MsgExitPool\\": MsgExitPool
};"
`;

exports[`toObjectWithEncodedMethods 1`] = `
"export const encoded = {
  joinPool(value: MsgJoinPool) {
    return {
      type_url: \\"/cosmos.pools.transfer.v1.MsgJoinPool\\",
      value: MsgJoinPool.encode(value).finish()
    };
  },

  exitPool(value: MsgExitPool) {
    return {
      type_url: \\"/cosmos.pools.transfer.v1.MsgExitPool\\",
      value: MsgExitPool.encode(value).finish()
    };
  }

};"
`;

exports[`toObjectWithPartialMethods 1`] = `
"export const messages = {
  joinPool(value: MsgJoinPool) {
    return {
      typeUrl: \\"/cosmos.pools.transfer.v1.MsgJoinPool\\",
      value: MsgJoinPool.fromPartial(value)
    };
  },

  exitPool(value: MsgExitPool) {
    return {
      typeUrl: \\"/cosmos.pools.transfer.v1.MsgExitPool\\",
      value: MsgExitPool.fromPartial(value)
    };
  }

};"
`;
